### PR TITLE
perf(rust, python): speed up write_csv for time-zone-aware columns

### PR DESCRIPTION
**Pull request by MarcoGorelli**
I'd expect this to improve performance, as it means that the time zone doesn't need to be parsed for each element, but I'm running a timing test to check, will post results when it's done

EDIT timing test on 2.5 million rows: https://www.kaggle.com/code/marcogorelli/polars-timing?scriptVersionId=131434660

here:
- 4.8 s ± 158 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
- 2.48 s ± 102 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

upstream/main:
- 5.19 s ± 671 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
- 2.57 s ± 168 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

### CODE REVIEW

File: polars/polars-io/src/csv/write_impl.rs [hunk 1]
```diff
@@ -293,11 +323,18 @@ pub(crate) fn write<W: Write>(
                 for (i, col) in &mut col_iters.iter_mut().enumerate() {
                     let datetime_format = match &options.datetime_format {
                         Some(datetime_format) => Some(datetime_format.as_str()),
```
Comment (ritchie46) [inline polars/polars-io/src/csv/write_impl.rs:325]: We could prepare this one outside of the while loop as well. Prepare a `Vec` with either `formats` or `datetime_format.as_str` for every column.

File: polars/polars-io/src/csv/write_impl.rs [hunk 2]
```diff
@@ -293,11 +323,18 @@ pub(crate) fn write<W: Write>(
...
-                        None => unsafe { *formats.as_ref().unwrap().get_unchecked(i) },
+                        None => unsafe { *datetime_formats.get_unchecked(i) },
                     };
+                    let time_zone = unsafe { *time_zones.get_unchecked(i) };
```
Comment (ritchie46) [inline polars/polars-io/src/csv/write_impl.rs:328]: Can we pass the `time_zones` slice to `write_anyvalue`. Most columns don't need to access the time_zone, so we don't need to take the value unless we write it.

If we also send the current index `i` to `write_anyvalue`. That function could determine itself if it should get the `time_zone` or not.

**Review by ritchie46 — COMMENTED** [review]
Thanks! Given that we are optimizing here, I have got a few other tips. :)

File: polars/polars-io/src/csv/write_impl.rs [hunk 3]
```diff
@@ -59,8 +59,9 @@ fn write_anyvalue(
     f: &mut Vec<u8>,
```
Comment (ritchie46) [inline polars/polars-io/src/csv/write_impl.rs:59]: We should mark the function as `unsafe` now.

File: polars/polars-io/src/csv/write_impl.rs [hunk 4]
```diff
@@ -129,7 +131,7 @@ fn write_anyvalue(
         #[cfg(feature = "dtype-datetime")]
         AnyValue::Datetime(_, _, tz) => {
             // If this is a datetime, then datetime_format was either set or inferred.
-            let datetime_format = datetime_format.unwrap_or_default();
+            let datetime_format = unsafe { *datetime_formats.get_unchecked(i) }.unwrap();
```
Comment (ritchie46) [inline polars/polars-io/src/csv/write_impl.rs:134]: This `unsafe` should be on the function.

If we `unwrap` the `Option<T>` in `datetime_formats`, why don't we fill it with `T` itself? That saves another branch in a hot loop.

**Comment by ritchie46** [comment]
Nice improvement @MarcoGorelli. :raised_hands:

**Comment by MarcoGorelli** [comment]
Thank you, review + tips very much appreciated!

**Pull request merged**